manifestoes/extensions.rst

```diff
@@ -0,0 +1,1675 @@
+  we will not implement it this way, we can imagine that GHC becomes
+  a set of programs that happen to share a binary; the choice of which
+  program is chosen by the language edition.
+
+- For backward compatibility, a language edition can be specified
+  at an arbitrary place in a command-line invocation of GHC, or in a
+  ``LANGUAGE`` pragma in a file. Even though it might come later in
+  a command line, the edition can affect the meaning of command-line
+  arguments that precede it.
+
+- A cabal file will allow a new
+  field ``language-edition``, available both at top-level and in
+  build-product stanzas. This will specify the language edition. To
+  support backward compatibility, this will use the ``default-language``
+  setting if that is available, and omitting the ``language-edition`` will
+  use the default. At some point, it is expected that ``language-edition``
+  will become required.
+
+- Once e.g. ``Stable2027`` is released, new language features will *not*
+  be available with the 2024 editions. That is, if we introduce a new
+  feature ``-XDependentTypes`` in 2028, then enabling ``-XDependentTypes``
+  with ``Stable2024`` (or even ``Experimental2024``) will be an error.
+  This policy gently encourages users to upgrade their editions in order
+  to access GHC's new features.
+
+- Once an edition has been eclipsed by newer models (that is, once the
+  calendar reads 2028 but someone is still using ``2024``), we know that
+  we are compiling older code. Our priority for such maintenance modes
+  is stability, not, say, making sure that the set of warnings conforms
+  to the latest standard.
+
+- We introduce several coarse-grained *semantic bundles* that group
+  together similar features. The individual choices are detailed below.
+  When using one of the new language
+  editions, error messages suggest opting into one of these bundles,
+  rather than suggesting individual extensions.
+
+- When printing out the namne of a warning flag as part of a warning,
+  we also include any bundle that also controls the warning.
+
+Detailed specification
+~~~~~~~~~~~~~~~~~~~~~~
+
+The chart below classifies all current extensions and warnings.
+
+* The GHC2021 column has an X for extensions enabled as part of GHC2021.
+  For warnings, it lists if the warning is part of a current warning bucket.
+
+* The next four columns describe ``Student2024``, ``Stable2024``, ``Experimental2024``, and ``Latest``,
+  respectively. Here is the key for these columns:
+
+  * Y: an extension is turned on
+
+  * blank: the feature is available, but not on
+
+  * N: trying to enable the feature is an error
+
+  * W: the warning warns by default
+
+  * E: the warning is an error by default
+
+  * W!: the warning warns by default and cannot be turned off
+
+  * E!: the warning is an error by default and cannot be turned off or made into a warning
+
+* The next several columns describe semantic bundles of options.
```

Comment:
I'm on the fence about these. My instinct is to say that I want to say something more like "FancyTypes but without the dubious stuff (for some personal definition of dubious)". But in practice I agree with the statement elsewhere that other languages tend to not give people this level of fine-grained control and it's fine: you just don't use the stuff you don't like.

Reply:
As a Haskeller, I would prefer the fine-grained control. I'll happily declare which features I want and use just those. I like tight specifications.

But I actively want Haskell to be easier to use by people unlike me. "Is that an existential type? or a rank-n type? Both have foralls in weird places..... Oh, forget it, I'll just do without."

The big counterpoint here is that an enterprise might want to choose just certain fancy types. But they still can, if they want to. Or at least as much as our extension system allows: one might reasonably be happy for datatypes to introduce new equalities but not want proper existentials, yet there's no way to say this.